include/swift/AST/Import.h

```diff
@@ -35,6 +43,282 @@ enum class ImportKind : uint8_t {
   Func
 };
 
+namespace detail {
+  using ImportPathElement = Located<Identifier>;
+  using ImportPathRaw = llvm::ArrayRef<ImportPathElement>;
+
+  template<typename Subclass>
+  class ImportPathBase {
+  public:
+    using Element = ImportPathElement;
+    using Raw = ImportPathRaw;
+
+  protected:
+    Raw raw;
+
+    ImportPathBase(Raw raw) : raw(raw) { }
+
+  public:
+    const Raw &getRaw() const { return raw; }
+
+    Raw::iterator begin() const {
+      return raw.begin();
+    }
+
+    Raw::iterator end() const {
+      return raw.end();
+    }
+
+    const Element &operator[](size_t i) const { return raw[i]; }
+    bool empty() const { return raw.empty(); }
+    size_t size() const { return raw.size(); }
+
+    const Element &front() const { return raw.front(); }
+    const Element &back() const { return raw.back(); }
+
+    /// True if \c this and \c other are precisely equal, including SourceLocs.
+    bool operator==(const Subclass &other) const {
+      return raw == other.raw;
+    }
+
+    /// True if \c this and \c other contain the same identifiers in the same
+    /// order, ignoring SourceLocs.
+    bool isSameAs(const Subclass &other) const {
+      return size() == other.size()
+             && std::equal(this->begin(), this->end(), other.begin(),
+                  [](const Element &l, const Element &r) -> bool {
+                    return l.Item == r.Item;
+                  }
+                );
+    }
+
+    Subclass takeTopLevel() const {
+      assert(size() >= 1 && "nothing to take");
+      return Subclass(raw.take_front());
+    }
+
+    Subclass takeParent() const {
+      assert(size() >= 0 && "nothing to take");
+      return Subclass(raw.drop_back());
+    }
```

Comment:
Nit: "take" usually indicates a consuming operation, perhaps these should be `getTopLevel` and `getParent`?

Reply:
Hmm, maybe. I was basing this on the underlying `take_front()`. 

(At one point, I had `topLevel()` instead of `front()`, but I changed my mind about this because it caused too much friction. `takeTopLevel()` was to contrast with `topLevel()`.

Reply:
I've changed these to `getTopLevelPath()` and `getParentPath()`.